docs/blog/2024-09-30-transactional-outbox.md

```diff
@@ -0,0 +1,170 @@
+
+We would rewrite our Spring Boot endpoint to only write a transaction to the database. The SQL for that transaction would look something like:
+
+```
+BEGIN TRANSACTION;
+INSERT INTO user VALUES ...;
+INSERT INTO outbox VALUES ...; # Insert the record for the queue
+COMMIT;
+```
+
+The Spring Boot application should have another thread which reads records from the `outbox` table, publishes them to the queue or streaming system, and updates the record in the database as "read".
+
+:::warning
+The topic of Exactly-Once Semantics is complex; we do not have time in this post to discuss the implications of EOS and a Transactional Outbox.
+
+As a hint, you can achieve EOS if you transactionally store the last-written offset inside your message broker. There are many "gotchas" to this depending on your message broker; for example, in Apache Kafka you must use `read_committed` consumers.
+:::
+
+### Using LittleHorse
+
+The Outbox Pattern is necessary to persist outgoing records in the case that we suffer a crash between writing to the database and writing to the record queue. However, what if we could "delegate" persistence and reliability to some other system?
+
+_Drumroll please..._
+
+Enter LittleHorse! What if we had a `WfSpec` that defined our process, as follows:
+
+```java
+public void wfLogic(WorkflowThread wf) {
+    var userRequest = wf.addVariable("create-user-request", JSON_OBJ).required();
+    wf.execute("create-user", userRequst);
+    wf.execute("send-welcome-email", user);
+}
+```
+Now, all our REST endpoint has to do is run the worklfow:
+
+```java
+@PostMapping("/user")
+public ResponseEntity<String> createUser(@RequestBody CreateUserRequest request) {
+    // Just run the workflow
+    littlehorsClient.runWf(RunWfRequest.newBuilder()
+        .setWfSpecName("user-workflow")
+        .putVariables("create-user-request", LHLibUtil.objToVarVal(request))
+        .build());
+    return ResponseEntity.status(HttpStatus.CREATED);
+}
+```
+
+No outbox table needed! If creating the user in the database fails, or if sending the welcome email fails, LittleHorse will patiently retry (according to your retry backoff policy) the `TaskRun`s until they succeed. In the event that you exhaust your retries, you still haven't lost data:
+
+* You can easily search for failed workflows.
+* You can restart failed workflows with the `rpc RescueThreadRun` once the database incident is resolved.
+
+:::tip
+You can add retries using the `Workflow` object:
+
+```java
+Workflow wf = Workflow.newWorkflow("user-workflow", this::wfLogic);
+wf.setDefaultTaskRetries(10);
+wf.setDefaultTaskExponentialBackoffPolicy(ExponentialBackoffRetryPolicy.newBuilder()
+        .setBaseIntervalMs(1000)
+        .setMultiplier(3)
+        .build());
+
+wf.registerWfSpec(littlehorseClient);
+```
+:::
+
+## Wrapping Up
```

Comment:
Maybe add those diagrams I sent you here in the wrap-up? Or their own section comparing the overall architectures.